Implement floating_modifier and mouse operations for floating views

RyanDwyer opened the pull request:
This implements the following:

* `floating_modifier` configuration directive (docs: https://i3wm.org/docs/userguide.html#floating_modifier)
* Drag a floating window by its title bar
* Hold mod + drag a floating window from anywhere
* Resize a floating view by dragging the border
* Resize a floating view by holding mod and right clicking anywhere on the view
* Resize a floating view and keep aspect ratio by holding shift while resizing using either method
* Mouse cursor turns into resize when hovering floating border or corner

Resizing *tiled* containers by dragging the border is not implemented in this PR.

Minimum and maximum bounds are enforced.

If the view being dragged or resized decides to unmap itself or do some other significant operation (like moving to another workspace or becoming fullscreen) then the mouse operation is cancelled via a call to `container_end_mouse_operation`. Keep an eye out for any places where I don't call this but should.

Unrelated to this PR, but note that views which have CSD/geo currently have damage issues, so maybe do your testing with an xwayland view.

Also I noticed I can't manually change a border's width using swaymsg any more (in master), so maybe change your border config before launching sway.

RyanDwyer commented on sway/input/cursor.c:
```diff
@@ -242,8 +464,79 @@ static void handle_cursor_motion_absolute(
+		seat->operation = OP_RESIZE;
+		seat->op_container = cont;
+		seat->op_resize_preserve_ratio = keyboard &&
+			(keyboard->modifiers.depressed & 1); // Shift
```
If anyone knows how I can replace `1` with a constant or more meaningful value, let me know.

progandy replied:
WLR_MODIFIER_SHIFT is defined in wlroots/include/wlr/types/wlr_keyboard.h. 

Is using the structure directly fine here or should you call wlr_keyboard_get_modifiers?

RyanDwyer replied:
I replaced the value - thanks.

I'm not sure what `wlr_keyboard_get_modifiers` is doing, so I'll wait for someone more knowledgeable to let me know if I should use it and what the consequences are of using the struct directly.

emersion replied:
`WLR_MODIFIER_*` constants cannot be used with `keyboard->modifiers.depressed`, they can only be used with `wlr_keyboard_get_modifiers`, because modifier indexes are not guaranteed to be the same from one keyboard layout to another.

SirCmpwn commented:
Running this as daily driver to test it, will report back with issues as I see them.

- Interactive resize with nautilus is awful, I think it's due to atomic update issues. Gedit is fine. Maybe if a client starts falling behind once or twice we should mark it as a known-slow client?
- If I float any GTK3 application and try to click near the right or bottom edges, it falls through to the tiled layer. I can also move my mouse up to the top or left edges, out of bounds, and still interact with the floating window. I think the geometry isn't being factored in here.

SirCmpwn commented:
- Resizing from the center is not how sway 0.15 nor i3 works, we should resize from the top left corner

SirCmpwn commented:
- Client initiated resize and move isn't working on GTK3 apps for me

RyanDwyer commented:
> Interactive resize with nautilus is awful, I think it's due to atomic update issues. Gedit is fine. Maybe if a client starts falling behind once or twice we should mark it as a known-slow client?

This works fine in rootston. I think we should be able to make it work even with atomic updates. I'll investigate this later.

> If I float any GTK3 application and try to click near the right or bottom edges, it falls through to the tiled layer. I can also move my mouse up to the top or left edges, out of bounds, and still interact with the floating window. I think the geometry isn't being factored in here.

Yeah, I didn't attempt to make it work with geometry because I know that's broken at the moment.

BTW, do we want to set a minimum grabbable area (eg. 5px) regardless of the border width?

> Resizing from the center is not how sway 0.15 nor i3 works, we should resize from the top left corner (using mod+drag)

Done. I don't think I've ever resized a view using mod + right click before this PR.

> Client initiated resize and move isn't working on GTK3 apps for me

Can you provide an example? I don't use GTK3 apps much, and I couldn't find any resize/move option in Nautilus.

emersion commented on sway/input/cursor.c:
```diff
@@ -194,15 +366,54 @@ void cursor_send_pointer_motion(struct sway_cursor *cursor, uint32_t time_msec,
+	} else if (c && container_is_floating(c)) {
+		// Try a floating container's resize edge
+		enum resize_edge edge = find_resize_edge(c, cursor);
+		if (edge == RESIZE_EDGE_NONE) {
```
This can be a `switch`, extracted in a separate function that returns a cursor image name from a resize edge.

emersion commented on sway/input/cursor.c:
```diff
@@ -242,8 +453,80 @@ static void handle_cursor_motion_absolute(
+
+	struct wlr_keyboard *keyboard = wlr_seat_get_keyboard(seat->wlr_seat);
+	bool mod_pressed = keyboard &&
+		(keyboard->modifiers.depressed & config->floating_mod);
```
Use `wlr_keyboard_get_modifiers`

RyanDwyer commented:
`request_move` and `request_resize` are now implemented.

Also note these related changes:

* Now uses `enum wlr_edges` instead of our own `enum resize_edge`
* Now uses `wlr_xcursor_get_resize_name` and removes our own `find_resize_edge_name`
* Renames `drag` to `move` for consistency

emersion commented on sway/commands/floating_modifier.c:
```diff
@@ -0,0 +1,30 @@
+#ifdef __linux__
```
We don't need all of these includes, do we?

emersion commented on sway/desktop/xdg_shell.c:
```diff
@@ -248,6 +253,24 @@ static void handle_request_fullscreen(struct wl_listener *listener, void *data)
+	struct sway_view *view = &xdg_shell_view->view;
+	struct wlr_xdg_toplevel_resize_event *e = data;
+	struct sway_seat *seat = e->seat->seat->data;
+	seat_begin_resize(seat, view->swayc, BTN_LEFT, e->edges);
```
The client can request a resize when it's not the left button that is pressed - iirc mpv requests resize on right click.

emersion replied:
We can probably use the serial to know which button the client is talking about.

emersion commented on sway/input/cursor.c:
```diff
@@ -139,13 +143,181 @@ static struct sway_container *container_at_coords(
+	desktop_damage_whole_container(con);
+}
+
+static void calculate_floating_constraints(struct sway_container *con,
```
Views themselves have size constraints. Should we handle these here?

RyanDwyer replied:
This is implemented BTW. See commit "Consider view's min/max sizes when resizing".

emersion commented on sway/input/cursor.c:
```diff
@@ -194,15 +366,27 @@ void cursor_send_pointer_motion(struct sway_cursor *cursor, uint32_t time_msec,
+		enum wlr_edges edge = find_resize_edge(c, cursor);
+		const char *image = edge == WLR_EDGE_NONE ?
+			"left_ptr" : wlr_xcursor_get_resize_name(edge);
+		wlr_xcursor_manager_set_cursor_image(cursor->xcursor_manager, image,
```
This will set the cursor each time the mouse moves. This is very inefficient as this re-uploads the image each time. We could probably save the current cursor image name in `sway_cursor`.

emersion commented on sway/input/seat.c:
```diff
@@ -888,3 +894,35 @@ struct seat_config *seat_get_config(struct sway_seat *seat) {
+	}
+	seat->operation = OP_MOVE;
+	seat->op_container = con;
+	seat->op_button = BTN_LEFT;
```
What if the client requested to move while another button was pressed?

emersion commented on sway/tree/container.c:
```diff
@@ -1103,3 +1107,13 @@ static bool find_urgent_iterator(struct sway_container *con,
+	struct sway_seat *seat;
+	wl_list_for_each(seat, &input_manager->seats, link) {
+		if (seat->op_container == container) {
+			seat->op_container = NULL;
```
Hmm, is this enough? Shouldn't this use `handle_end_operation`?

emersion commented:
* You can drag tiled windows (open `gnome-calculator`, drag the titlebar, move around, release - it segfaults)
* Damage tracking for CSD windows is messed up - but that's a window geometry issue, so it doesn't belong to this PR (open `gnome-calculator` preferences, drag it around)
* Xwayland's move/resize requests are ignored (`wlr_xwayland_surface.request_{move,resize}`): try `env WAYLAND_DISPLAY= gnome-calculator` and try dragging the titlebar
* I have damage issues when resizing (shrinking) mpv under xwayland
* For some reason the cursor is not updated when I resize or move - it sticks to the arrow cursor

RyanDwyer commented:
> You can drag tiled windows (open gnome-calculator, drag the titlebar, move around, release - it segfaults)

Fixed.

> Xwayland's move/resize requests are ignored (wlr_xwayland_surface.request_{move,resize}): try env WAYLAND_DISPLAY= gnome-calculator and try dragging the titlebar

Implemented now. I noticed you can right click the title bar in xwayland gnome-calculator and there are resize and move options there, but clicking these don't fire the request_resize/move functions so I'm not sure sure how that's supposed to work.

> I have damage issues when resizing (shrinking) mpv under xwayland

Can't replicate.

> For some reason the cursor is not updated when I resize or move - it sticks to the arrow cursor

Also can't replicate.

I also fixed a bug where if you had a tiled container behind a floating container, then clicked the tiled container then clicked the floating container, focus was not given to the floating container properly.

emersion commented:
>I have damage issues when resizing (shrinking) mpv under xwayland

Can replicate on both the DRM backend and the Wayland backend.

![screenshot](https://user-images.githubusercontent.com/506932/43045608-bf39e6a4-8db3-11e8-89d1-187d230374f0.png)

>For some reason the cursor is not updated when I resize or move - it sticks to the arrow cursor

Same.

emersion commented:
Found why the cursor isn't updated: there's a `if (seat->operation != OP_NONE)` that returns immediately in `cursor_send_pointer_motion` so the "Handle cursor image" part is never reached when resizing/moving.

emersion commented:
`seat_begin_move` and `seat_begin_resize` should maybe set the correct cursor?

SirCmpwn commented:
\o/

ben-foxmoore commented:
> Resizing from the center is not how sway 0.15 nor i3 works, we should resize from the top left corner (using mod+drag)

Resizing in sway 0.15 was dependent on which quarter of the window the cursor started in. The corner furthest from the cursor is locked in place. I.e. it was possible to drag the top left corner of the window towards the top left corner of the screen.

With the current implementation, you have to drag the bottom right corner, and then move the window. Is this an intentional change in functionality?

RyanDwyer commented:
Not intentional. I'll add it to my todo list.

InfernoZeus commented:
>Resizing from the center is not how sway 0.15 nor i3 works, we should resize from the top left corner (using mod+drag)

Resizing in sway 0.15 was dependent on which quarter of the window the cursor started in. The corner furthest from the cursor is locked in place. I.e. it was possible to drag the top left corner of the window towards the top left corner of the screen.

With the current implementation, you have to drag the bottom right corner, and then move the window. Is this an intentional change in functionality?

Edit: sorry, reposted on my personal account, so message is out of order.
